### ISSUE TITLE
Ignore `catch` block in the `no-empty` rule

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#2808]
## Problem

The following is very common pattern when you don't care about an exception:

```js
try {
  foo();
} catch (err) {}
```

But it's caught by the [`no-empty` rule](http://eslint.org/docs/rules/no-empty).

I want the rule to catch empty blocks, but ignore the `catch` block as it's empty because it's required by JS, not because I want it there.

I know I can put in a `// empty` comment there, but that feels like an unuseful reiteration of the obvious.

## Possible solution 1

Ignore `catch` by default as it's required to be empty and not really useful to require it not to be empty. This is the preferred solution.

## Possible solution 2

Add an option that defines exceptions in the form of keywords:

```json
{
  "no-empty": [2, {"exceptions": "catch"}]
}
```

**Comment by nzakas** [comment]
This has already been discussed at length here: https://github.com/eslint/eslint/issues/1841

tl;dr The rule used to omit catch by default, but then people were missing empty catches that should not have been empty and there was no way to differentiate. The best solution was to use a comment to mark the ones meant to be empty.

**Comment by sindresorhus** [comment]
From reading that thread it seemed like everyone disagreed with your position. I realize it's not a good default, but I thought ESLint favored configurability over being opinionated. This decision smells a lot like the latter. I don't understand why a simple `exceptions` option is unacceptable.

The result of this is that the `no-empty` rule is useless for most user.

-

*Sorry about not finding that ticket. I searched the issue tracker for a while before opening this...*

**Comment by ilyavolodin** [comment]
@sindresorhus We provide a way to customize this behavior, if I'm not mistaken this code will lint just fine
```js
try {
    foo();
} catch(err) {
    // I want application to not crush, but don't care about the message
}
```
Any comment inside a block will mark it as non-empty. Leaving comments in the cases where you used try catch but ignoring cache is a good idea, if somebody else is going to be looking at your code.
But in general, I do see your point.

**Comment by sindresorhus** [comment]
@ilyavolodin I'm aware, as already explained in the issue description, but I don't want to add an IMHO useless comment just to please ESLint.

**Comment by nzakas** [comment]
@sindresorhus 

> From reading that thread it seemed like everyone disagreed with your position. I realize it's not a good default, but I thought ESLint favored configurability over being opinionated. This decision smells a lot like the latter. I don't understand why a simple exceptions option is unacceptable.

Being unopinionated doesn't mean we automatically do everything everyone wants. It means we don't prescribe styles or force you to do anything you don't want to do. We also make no promises that your code will lint without any changes or that all of our rules are appropriate for all code. (Also discussed: https://github.com/eslint/eslint/issues/1841#issuecomment-87797166)

Here, we have provided a rule that you can choose to use or not use, and we have several ways of having one-off disabling of its warnings, including putting a comment in the block (an approach we use in other rules as well) or use eslint-disable comments. You also have the option of writing your own rule to use instead of this one.

The problem with adding an exception (and the exception we had at one time) is that there's no way to distinguish intentional and unintentional empty catch blocks. This is a far more dangerous circumstance than any other, as swallowing errors can hide significant problems. I don't like having an option that is this much of a landmine for developers, because I can tell you exactly what will happen: someone will miss a swallowed error using this exception and then they will come back and ask for a way to opt out of the exception somehow. In which case, we are now back to square one where we need comments to indicate which catch blocks should be omitted. This is exactly how #1841 came about.

This is part of our challenge, sometimes pure static analysis isn't good enough to figure out intent and we need to do something to augment it.

At this point, I don't see a good argument for adding an option that could result in false negatives. If you have an idea for solving the false negative problem with the exception you're requesting, I'll gladly consider it.

**Comment by sindresorhus** [comment]
> I don't like having an option that is this much of a landmine for developers, because I can tell you exactly what will happen: someone will miss a swallowed error using this exception

The dev can just as easily disable the whole rule. Same result. The dev is in control of both. If you start thinking like this pretty much all the exceptions in the other rules are a mistake. Even ability to disable rules would be a mistake with that logic.

But it's clear we're seeing things differently, so I'm going to stop here, disable the rule, and look into making a plugin :)

**Comment by xjamundx** [comment]
@sindresorhus https://github.com/alex-shnayder/eslint-plugin-no-empty-blocks :)

**Comment by timkendrick** [comment]
FYI, this decision is also causing issues when using the babel-eslint parser, which strips out these comments as part of its compilation process (see babel/babel-eslint#141, babel/babel#1904, babel/babel#758).

It seems a shame that we need to rely on third-party plugins to get this feature working, but as @sindresorhus says, it's clear we're seeing things differently.

**Comment by nzakas** [comment]
Or you could ask babel to not strip those comments. :)

**Comment by aliak00** [comment]
jslint has an IMO nice approach where if you name the variable `ignore`, well, then, it's ignored. And it complains if the variable named `ignore` is used in a catch block. I'd assume it comes with the option to disable/enable I guess. Would that work for eslint?

**Comment by michaelficarra** [comment]
@aliak00 I like that. Can you open a new issue for it?

**Comment by aliak00** [comment]
@michaelficarra Heya! Issue creation guide page say implementation is my responsibility. But I don't think I'd want to implement this. It was just a suggestion I threw in as a compromise instead of globally turning off empty catch blocks, which I personally don't think is a good idea :) - and which I see has gone in with #5800 - and I frankly don't mind just adding a `// eslint-disable-line no-empty` near the quite rarely valid (IMO) empty catch block cases. But of course feel free to open an issue for this!

**Comment by cmmartin** [comment]
+1 for globally turning off empty `catch` blocks. There is nothing wrong with an empty `catch` block and the chances someone just plain forgets to implement it are infinitely less likely to occur than the very common case of wanting an empty `catch` block. This is overreach.

**Comment by platinumazure** [comment]
@cmmartin Please use the [`allowEmptyCatch` option](http://eslint.org/docs/rules/no-empty#options) to globally turn off the rule for empty catch blocks. Thanks!

**Comment by danny-andrews** [comment]
@sindresorhus 
> The following is very common pattern when you don't care about an exception:
```js
try {
  foo();
} catch (err) {}
```

This is a HORRIBLE, HORRIBLE pattern that, unfortunately, is very common in javascript. But it shouldn't be perpetuated explicitly or implicitly. At the **very least** you should provide a comment explaining why you are ignoring that exception. Better yet, you should do an `instanceof` check to only ignore `Error`s of a certain type, and rethrow all others, which would also satisfy this rule.

p.s. Excuse me for the strong formatting, but good error handling is something I feel **VERY STRONGLY** about. :D
